compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2314,16 +2312,25 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       untpd.Block(makeCanThrow(capabilityProof), expr)
 
   def typedTry(tree: untpd.Try, pt: Type)(using Context): Try = {
+    // We want to type check tree.expr first to comput NotNullInfo, but `addCanThrowCapabilities`
+    // uses the types of patterns in `tree.cases` to determine the capabilities.
+    // Hence, we create a copy of cases with empty body and type check that first, then type check
+    // the rest of the tree in order.
+    val casesEmptyBody1 = tree.cases.mapconserve(cpy.CaseDef(_)(body = EmptyTree))
+    val casesEmptyBody2 = typedCases(casesEmptyBody1, EmptyTree, defn.ThrowableType, WildcardType)
```

Comment:
Does this work if the pattern uses a variable whose nullability is changed by the try block (`expr`)? Won't we be typing such a variable in the pattern with the wrong type here?